### ISSUE TITLE
Use search queries / list of IPs instead of single host field

### ISSUE THREAD

**Issue by zhelyan** [melezhik/chef-plugin#10]
I think this plugin will be significantly more useful if instead of a single hosts you can specify Chef search query or a list of fqdns/ips. They should both accept expressions (tokens, e.g using the Token Macro plugin)
You will be be then able to get a list of servers using e.g Scriptler script, save the list in a variable and pass this to the chef-plugin.
Alternatively, if the servers are already registered with a Chef server you will be able to use a search query to determine the targets.

**Comment by melezhik** [comment]
Hi Zhelyan! First of all  thanks for suggestion.

I still have some concerns about this.
- you are talking about ability to run chef-client plugin against more then one hosts, ok that is fine, but this is probably will complicate existed design and it is probably against some initial intentions of this plugin creation
- how you are going to handle different ssh credentials, different chef client configuration paths foe different host ?
- if you are going to use knife ssh command approach where ssh credentials for all nodes are the same (  at least  conventionally ) - then this plugin needs  to be changed to accept one host / knife ssh way to run, which is  again make things over complicated  - as both methods have different interfaces, set of parameters, etc. ...  and finally it's probably better have two different plugins - chef-client plugin and knife-ssh plugin for this.
- as summary I am not against of extension to accept  new features but _sometimes_ it is much easier to have many simple dedicated tools rather than have a single all-in-one hard to maintain software ...

**Comment by zhelyan** [comment]
Thank you for your comments,

Regarding providing different ssh credentials / client config (point 2) - the path to the ssh credentials can be passed as a build parameter/env variable to knife:
 `knife ssh ..... -i $IDENTITY_FILE`

Alternatively one can use e.g https://wiki.jenkins-ci.org/display/JENKINS/SSH+plugin

As for the the client configuration being used - again this could be an env variable or managed by e.g 
https://wiki.jenkins-ci.org/display/JENKINS/Config+File+Provider+Plugin

Point 3 - yes, the interfaces will be different but you can keep the existing functionality and provide i.e. `optionalBlock` to switch between simple, single host and advanced, query mode ?

My thinking is that nowadays servers are almost always managed as a group. Whether this group is an environment or a subset of one is irrelevant.
Because of this servers belonging to a group are often identified by a dynamic query which includes the group, not the servers e.g - `aws ec2 describe-instances --filter Name=tag:Project,Values=MyProject .....`, `knife search node 'chef_environment:prod'....` etc.

What I propose is that this plugin implements the latter ( querying the Chef server ) or the ability to specify multiple IP addresses in the text field so people can query Chef/Aws etc themselves and pass the results (ip addresses) to this plugin.

**Comment by melezhik** [comment]
Hi  Zhelyan!

Thanks for your comments. Probably you missing some my points

> knife ssh ..... -i $IDENTITY_FILE
>  this implies that all the hosts has the same $IDENTITY_FILE. Ok, this could be the case when we have homogeneous groups of hosts ( aws or whatever  ) which are managed similarly .
> But my use case what if we have host1 with one indemnity file and second with another identity files, then you need provide 4 parameters for  this request succeed.

Again, in _most_ cases  if we are given aws clouds or other groups of hosts orchestrated by single manager - it mostly always not the issue to provide ssh credentials for them, as they with the high chances there will be the same credentials for all the nodes in a group. But I am talking about other still real life situations.

Converging groups of hosts are fine, and I really got your point, but want  I try to say that it's probably there are better tools for it. ( knife ssh, ansible ). And we end up with something you already mentioned.

Currently jenkins chef-plugin is more look like "snowflake" tool. When you have a few hosts which should be setup specifically ( ssh user , ssh password or ssh identity file or none default path to chef client )  where this plugin could be handy 

Regards.

Alexey

**Comment by zhelyan** [comment]
Fair point! Still, how about updating the field to take more than 1 IP ?
Yes, the assumption is that all servers will use the same SSH credentials. The benefit is that you will be able to build the list of servers in e.g groovy, knife search, take it from build parameter etc and pass it on to the Run chef-client build step. The change to the interface is non-breaking and minimal ?

**Comment by melezhik** [comment]
Hi! I need to think about this a bit ...
